drivers: kscan: Add CAP1203 driver

keiya-nobuta opened the pull request:
Add driver for Microchip CAP1203 3-Channel Capacitive Touch Sensor

datasheet: https://ww1.microchip.com/downloads/en/DeviceDoc/00001572B.pdf


My testing environment:
- Board: Nucleo L031K6
- Sensor Board: [SparkFun Capacitive Touch Slider](https://www.sparkfun.com/products/15344)
- App: samples/drivers/kscan_touch

DeviceTree Sample:
```
&i2c1 {
	...snip...
	cap1203: cap1203@28 {
		compatible = "microchip,cap1203";
		reg = <0x28>;
		int-gpios = <&gpioa 12 GPIO_ACTIVE_LOW>;
		label = "CAP1203";
		status = "okay";
	};
};
```

keiya-nobuta commented:
I was mistaken, added copyright.

gmarull commented on drivers/kscan/Kconfig.cap1203:
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2022 Keiya Nobuta
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig KSCAN_CAP1203
+	bool "CAP1203 3-cannel capacitive touch sensor driver"
+	depends on I2C
```
```suggestion
DT_COMPAT_MICROCHIP_CAP1203 := microchip,cap1203

menuconfig KSCAN_CAP1203
	bool "CAP1203 3-cannel capacitive touch sensor driver"
	depends on I2C
	default $(dt_compat_enabled,$(DT_COMPAT_MICROCHIP_CAP1203))
```

gmarull reviewed (commented):
lgtm, one minor suggestion

keiya-nobuta commented:
Update:
- Apply @gmarull's suggestion.
- Fix `Boolean option 'KSCAN_CAP1203_INTERRUPT' prompt must not start with 'Enable...'` error.

mbolivar-nordic commented on drivers/kscan/kscan_cap1203.c:
```diff
@@ -0,0 +1,256 @@
+#define CAP1203_INIT(index)							\
+	static const struct cap1203_config cap1203_config_##index = {		\
+		.i2c = I2C_DT_SPEC_INST_GET(index),				\
+		IF_ENABLED(CONFIG_KSCAN_CAP1203_INTERRUPT,			\
+		(.int_gpio = GPIO_DT_SPEC_INST_GET(index, int_gpios),))		\
```
There is a bug here. If there are 2 nodes, and one has `int-gpios`, but the other does not, and CONFIG_KSCAN_CAP1203_INTERRUPT=y, I think this should be using GPIO_DT_SPEC_INST_GET_OR() instead, to set a `NULL` port value in the spec. Then you can just skip devices without interrupt GPIOs, right?

I didn't review in detail though, so not going to block or approve for now.

mbolivar-nordic replied:
Yes, that was a big vague, sorry. I mean something like this here:

```c
.int_gpio = GPIO_DT_SPEC_INST_GET(index, int_gpios, {0}),
```

Then, in the initialization routine:

```c
if (int_gpio.port != NULL) { do_something_with(&int_gpio); }
```

I.e. we skip the block of code that does something with the interrupt GPIO if there isn't one to do anything with.

> I don't know if this is a bug, but if so, this driver is implemented by imitating [drivers/kscan/kscan_ft3506.c](https://github.com/zephyrproject-rtos/zephyr/blob/70c48417c7a0d7db139accc2368bec7f4aa9c762/drivers/kscan/kscan_ft5336.c#L231-L232), so there may be a bug in kscan_ft5336 as well.

I'm starting to notice this pattern in other reviews as well. From the devicetree side of things, the property is not required, so I think that in the general case, drivers should generally be able to handle devices without such GPIOs even when interrupt support is enabled in software. But if there is something hardware-specific I am missing here, please let me know. Thanks.

keiya-nobuta replied:
Thanks, I understand it.

In the current implementation, only interrupt mode or polling mode generates input events from the touch sensor, and this can be switched with Kconfig. Therefore, if interrupt mode is set and there is no int-gpio, the device will not be working.

I think it's okay to switch between using interrupt mode and polling mode depending on the presence or absence of int-gpio, so I'll change it. (However, I think the footprint will be larger.)

keiya-nobuta commented:
I changed the following:
- Interrupt mode is default used when int-gpios is set on dts.
- Polling mode is optional and it is used when int-gpios is not set.

gmarull commented on drivers/kscan/kscan_cap1203.c:
```diff
@@ -0,0 +1,267 @@
+
+#define DT_DRV_COMPAT microchip_cap1203
+
+#include <drivers/kscan.h>
+#include <drivers/i2c.h>
+#include <drivers/gpio.h>
+
+#include <logging/log.h>
```
```suggestion
#include <zephyr/drivers/kscan.h>
#include <zephyr/drivers/i2c.h>
#include <zephyr/drivers/gpio.h>

#include <zephyr/logging/log.h>
```

keiya-nobuta replied:
OK, I'll change these.

gmarull commented on drivers/kscan/kscan_cap1203.c:
```diff
@@ -0,0 +1,267 @@
+{
+	struct cap1203_data *data = dev->data;
+
+	const struct cap1203_config *config = dev->config;
+
+	if (config->int_gpio.port != NULL) {
+		const struct cap1203_config *config = dev->config;
```
duplicate config here

keiya-nobuta replied:
I'll fix it.

gmarull commented on drivers/kscan/kscan_cap1203.c:
```diff
@@ -0,0 +1,267 @@
+{
+	struct cap1203_data *data = dev->data;
+
+	const struct cap1203_config *config = dev->config;
+
+	if (config->int_gpio.port != NULL) {
+		const struct cap1203_config *config = dev->config;
```
ditto

keiya-nobuta replied:
I'll fix it.

gmarull commented on drivers/kscan/kscan_cap1203.c:
```diff
@@ -0,0 +1,267 @@
+
+	k_work_init(&data->work, cap1203_work_handler);
+
+	if (config->int_gpio.port != NULL) {
+		if (!device_is_ready(config->int_gpio.port)) {
```
```suggestion
	if (config->int_gpio.port != NULL) {
		int r;

		if (!device_is_ready(config->int_gpio.port)) {
```

keiya-nobuta replied:
I wrote below, I'll change it according to the addition of the interrupt disabling at init time.

gmarull commented on drivers/kscan/kscan_cap1203.c:
```diff
@@ -0,0 +1,267 @@
+{
+	const struct cap1203_config *config = dev->config;
+	struct cap1203_data *data = dev->data;
+	int r = -ENOTSUP;
```
```suggestion
```

keiya-nobuta replied:
OK

gmarull commented on drivers/kscan/kscan_cap1203.c:
```diff
@@ -0,0 +1,267 @@
+#ifdef CONFIG_KSCAN_CAP1203_POLL
+	else {
+		k_timer_init(&data->timer, cap1203_timer_handler, NULL);
+		r = 0;
```
```suggestion
```

keiya-nobuta replied:
OK

gmarull commented on drivers/kscan/kscan_cap1203.c:
```diff
@@ -0,0 +1,267 @@
+	}
+#endif
+
+	if (r < 0) {
+		LOG_ERR("Could not configure device");
+	}
+
+	return r;
```
this path seems unfeasible, as you always log error and return directly, just`return 0` at the end.

```suggestion
	return 0;
```

keiya-nobuta replied:
OK, I'll fix it.

gmarull commented on drivers/kscan/kscan_cap1203.c:
```diff
@@ -0,0 +1,267 @@
+	const struct cap1203_config *config = dev->config;
+	int r;
+
+	if (!callback) {
+		LOG_ERR("Invalid callback (NULL)");
+		return -EINVAL;
+	}
```
no need to protect against programming errors

```suggestion
```

keiya-nobuta replied:
OK

gmarull commented on drivers/kscan/kscan_cap1203.c:
```diff
@@ -0,0 +1,267 @@
+	return i2c_reg_write_byte_dt(i2c, REG_MAIN_CONTROL, ctrl);
+}
+
+static int cap1203_enable_interrupt(const struct i2c_dt_spec *i2c, uint8_t enable)
```
please use booleans for boolean flags
```suggestion
static int cap1203_enable_interrupt(const struct i2c_dt_spec *i2c, bool enable)
```

keiya-nobuta replied:
The value to be written to the register is specified here, but I think that boolean is more appropriate, so I'll change it.

gmarull commented on drivers/kscan/kscan_cap1203.c:
```diff
@@ -0,0 +1,267 @@
+		}
+
+		r = cap1203_enable_interrupt(&config->i2c, INTERRUPT_ENABLE);
+	} else {
+		r = cap1203_enable_interrupt(&config->i2c, INTERRUPT_DISABLE);
+	}
```
interrupt can be disabled once at init time, since int_gpio will be NULL all the time. why doing it every time?

keiya-nobuta replied:
interrupt disabling doesn't have to do it every time. I'll change to do it at init time.

keiya-nobuta committed:
drivers: kscan: Add CAP1203 driver

Add driver for Microchip CAP1203 3-Channel Capacitive Touch Sensor

datasheet: https://ww1.microchip.com/downloads/en/DeviceDoc/00001572B.pdf

Signed-off-by: Keiya Nobuta <nobuta.keiya@fujitsu.com>

keiya-nobuta commented:
I changed the following:
- fixed the patch according to @gmarull's suggestions.
- (and rebase from main)
